enhancements/windows-containers/windows-machine-config-operator-upgrades.md

```diff
@@ -0,0 +1,185 @@
+will then be able to configure VMs that will be created in place of the
+terminated ones.
+
+In order to allow for this to occur with minimal service disruption, it is
+recommended that users have at least two Windows nodes within the cluster.
+
+All Windows Kubernetes component updates will be tied to WMCO releases.
+
+### Justification
+
+The reason we need to cause the VMs to be fully terminated, instead of updating
+them, is that changes made to the VM done in previous versions of the WMCO may
+cause issues with the current version. If we chose to do an in-place upgrade,
+we would need to actively undo changes made to the node that may cause issues
+with the current WMCO version. The logic to keep track and undo these changes
+adds an additional layer of complexity that also adds the benefit of a quicker
+upgrade time. However, in-place upgrades do not provide any key benefits
+for Windows workloads. For example, Windows has [restrictions regarding storage](https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#storage)
+preventing the projection of volumes from host storage, because of this we do
+not have to worry about preserving node local data.
+With all aspects weighed, the risk of introducing additional issues to provide
+an in-place upgrade is not worth it at this time, however it should be explored
+as the product matures.
+
+The reason we will not handle Windows operating system updates is that the
+cluster administrator provides the Windows image to create the VMs with, and
+it stays their responsiblity to provide an updated image.
+
+### Design Details
+
+WMCO is published to OperatorHub as a Red Hat operator. When a new WMCO version
+is released, and the new version is compatible with the current cluster
+version, an update can occur. While we cannot explicity specify a required OCP
+version for the WMCO upgrade, each new OCP version upgrade should include a
+Kubernetes version upgrade. This allows us to use the [minKubeVersion](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/building-your-csv.md#operator-metadata)
+field in WMCO's CSV, which stops the operator from upgrading if the next
+WMCO version does not have at least the specified Kubernetes version.
+
+This update will either happen automatically, or require approval, based on
+the settings given through OLM when initially installing the operator.
+
+In order to facilitate upgrades, WMCO will add an annotation to the node
+indicating the version of the operator which configured the node.
+Using a node annotation is the easiest way to track this information in a way
+that lasts across WMCO restarts. WMCO is directly responsible for the creation
+of Windows nodes, and we should expect that other entities will not clear or
+alter the annotations WMCO gives to such nodes.
+
+On reconcile, if the annotated version does not match the current version,
```

Comment:
There are a couple elements missing from this implementation.

Not all clusters are running the machine-api, so that is one thing we should call out in this design.  If we're only supporting clusters with the machine-api, we should state it.

Next, not all machines are backed by a machine-set.  The machine-set is the thing that ensures you get a new machine, similar to a pod and replica set relationship.  If we're requiring all Windows machines to be backed by a machine-set, we should specify that explicitly here.  Otherwise, we need to specify that we'll determine if the machine is owned by a machine-set and what to do if it's not.

Reply:
That is a good point, I have added an additional non-goal, as well as another paragraph below this one to make that clear.

Reply:
Yes, today, we require machine-api and we're only managing windows nodes that are part of a defined machine set. Agreed we should call that out.